extensions/gitpod/src/authentication.ts

```diff
@@ -0,0 +1,253 @@
+		this._register(vscode.authentication.registerAuthenticationProvider('gitpod', 'Gitpod', this, { supportsMultipleAccounts: false }));
+		this._register(this.context.secrets.onDidChange(() => this.checkForUpdates()));
+	}
+
+	get onDidChangeSessions() {
+		return this._sessionChangeEmitter.event;
+	}
+
+	async getSessions(scopes?: string[]): Promise<vscode.AuthenticationSession[]> {
+		// For the Gitpod scope list, order doesn't matter so we immediately sort the scopes
+		const sortedScopes = scopes?.sort() || [];
+		this._logger.info(`Getting sessions for ${sortedScopes.length ? sortedScopes.join(',') : 'all scopes'}...`);
+		const sessions = await this._sessionsPromise;
+		const finalSessions = sortedScopes.length
+			? sessions.filter(session => arrayEquals([...session.scopes].sort(), sortedScopes))
+			: sessions;
+
+		this._logger.info(`Got ${finalSessions.length} sessions for ${sortedScopes?.join(',') ?? 'all scopes'}...`);
+		return finalSessions;
+	}
+
+	private async checkForUpdates() {
+		const previousSessions = await this._sessionsPromise;
+		this._sessionsPromise = this.readSessions();
+		const storedSessions = await this._sessionsPromise;
+
+		const added: vscode.AuthenticationSession[] = [];
+		const removed: vscode.AuthenticationSession[] = [];
+
+		storedSessions.forEach(session => {
+			const matchesExisting = previousSessions.some(s => s.id === session.id);
+			// Another window added a session to the keychain, add it to our state as well
+			if (!matchesExisting) {
+				this._logger.info('Adding session found in keychain');
+				added.push(session);
+			}
+		});
+
+		previousSessions.forEach(session => {
+			const matchesExisting = storedSessions.some(s => s.id === session.id);
+			// Another window has logged out, remove from our state
+			if (!matchesExisting) {
+				this._logger.info('Removing session no longer found in keychain');
+				removed.push(session);
+			}
+		});
+
+		if (added.length || removed.length) {
+			this._sessionChangeEmitter.fire({ added, removed, changed: [] });
+		}
+	}
+
+	private async readSessions(): Promise<vscode.AuthenticationSession[]> {
+		let sessionData: SessionData[];
+		try {
+			this._logger.info('Reading sessions from keychain...');
+			const storedSessions = await this._keychain.getToken();
+			if (!storedSessions) {
+				return [];
+			}
+			this._logger.info('Got stored sessions!');
+
+			try {
+				sessionData = JSON.parse(storedSessions);
+			} catch (e) {
+				await this._keychain.deleteToken();
+				throw e;
+			}
+		} catch (e) {
+			this._logger.error(`Error reading token: ${e}`);
+			return [];
+		}
+
+		const sessionPromises = sessionData.map(async (session: SessionData) => {
+			// For the Gitpod scope list, order doesn't matter so we immediately sort the scopes
+			const sortedScopes = session.scopes.sort();
+			const scopesStr = sortedScopes.join(' ');
+
+			let userInfo: { id: string; accountName: string } | undefined;
+			if (!session.account) {
```

Comment:
We need to validate here that token still has scopes which are required by current version of extension. It will bit as next time when we add a new scope, but functionallity won't work since old token will be used without invalidation. It should be common pattern in all our clients.

Reply:
We don't need to validate scopes, scopes and token are 1 to 1 and stored in db, if we add/remove a scope in a new version of the extension then `getSessions` will be called with the new required scopes and it will filter from all the sessions returning only the ones which meets the required scopes.